File: aster-crypto/src/decrypt.rs
//
// Aster Communications Inc.
//
// Copyright (c) 2026 Aster Communications Inc.
//
// This file is part of this project.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the AGPLv3 as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// AGPLv3 for more details.
//
// You should have received a copy of the AGPLv3
// along with this program. If not, see <https://www.gnu.org/licenses/>.
//
use pgp::composed::{Deserializable, Message};

use chrono::Utc;

use crate::error::{CryptoError, Result};
use crate::keys::{KeyPair, PublicKey, PublicKeyInner};
use crate::sign::{
    signed_public_key_can_sign, signed_secret_key_can_sign, signed_secret_signing_identity_valid,
    signing_identity_valid,
};

pub const MAX_CIPHERTEXT_BYTES: usize = 16 * 1024 * 1024;
pub const MAX_PLAINTEXT_BYTES: usize = 64 * 1024 * 1024;

fn capped_plaintext(data: &[u8]) -> Result<Vec<u8>> {
    if data.len() > MAX_PLAINTEXT_BYTES {
        return Err(CryptoError::Decryption("Decryption failed".into()));
    }
    Ok(data.to_vec())
}

pub fn decrypt_message(ciphertext: &[u8], secret_keys: &[&KeyPair]) -> Result<Vec<u8>> {
    if secret_keys.is_empty() {
        return Err(CryptoError::KeyNotFound("No secret keys provided".into()));
    }
    if ciphertext.len() > MAX_CIPHERTEXT_BYTES {
        return Err(CryptoError::Decryption("Decryption failed".into()));
    }

    let (msg, _) = Message::from_armor_single(ciphertext)
        .map_err(|_| CryptoError::Decryption("Decryption failed".into()))?;

    for keypair in secret_keys {
        let decrypted = msg.decrypt(|| "".to_string(), &[keypair.secret_key()]);

        if let Ok((decrypted_msg, _key_ids)) = decrypted {
            if let Some(literal) = decrypted_msg.get_literal() {
                return capped_plaintext(literal.data());
            }

            if let Message::Literal(lit) = decrypted_msg {
                return capped_plaintext(lit.data());
            }
        }
    }

    Err(CryptoError::Decryption("Decryption failed".into()))
}

pub fn decrypt_message_binary(ciphertext: &[u8], secret_keys: &[&KeyPair]) -> Result<Vec<u8>> {
    if secret_keys.is_empty() {
        return Err(CryptoError::KeyNotFound("No secret keys provided".into()));
    }
    if ciphertext.len() > MAX_CIPHERTEXT_BYTES {
        return Err(CryptoError::Decryption("Decryption failed".into()));
    }

    let msg = Message::from_bytes(ciphertext)
        .map_err(|_| CryptoError::Decryption("Decryption failed".into()))?;

    for keypair in secret_keys {
        let decrypted = msg.decrypt(|| "".to_string(), &[keypair.secret_key()]);

        if let Ok((decrypted_msg, _key_ids)) = decrypted {
            if let Some(literal) = decrypted_msg.get_literal() {
                return capped_plaintext(literal.data());
            }

            if let Message::Literal(lit) = decrypted_msg {
                return capped_plaintext(lit.data());
            }
        }
    }

    Err(CryptoError::Decryption("Decryption failed".into()))
}

fn verify_with_sender_keys(msg: &Message, sender_keys: &[&PublicKey]) -> bool {
    let now = Utc::now();
    for pk in sender_keys {
        let verified = match &pk.inner {
            PublicKeyInner::Standalone(spk) => {
                signing_identity_valid(spk, now)
                    && signed_public_key_can_sign(spk)
                    && msg.verify(spk).is_ok()
            }
            PublicKeyInner::FromSecret(ssk) => {
                signed_secret_signing_identity_valid(ssk, now)
                    && signed_secret_key_can_sign(ssk)
                    && msg.verify(ssk).is_ok()
            }
        };
        if verified {
            return true;
        }
    }
    false
}

pub fn decrypt_and_verify(
    ciphertext: &[u8],
    secret_keys: &[&KeyPair],
    sender_keys: &[&PublicKey],
) -> Result<Vec<u8>> {
    if secret_keys.is_empty() {
        return Err(CryptoError::KeyNotFound("No secret keys provided".into()));
    }
    if sender_keys.is_empty() {
        return Err(CryptoError::SignatureVerification(
            "decrypt_and_verify requires at least one sender key".into(),
        ));
    }
    if ciphertext.len() > MAX_CIPHERTEXT_BYTES {
        return Err(CryptoError::Decryption("Decryption failed".into()));
    }

    let (msg, _) = Message::from_armor_single(ciphertext)
        .map_err(|_| CryptoError::Decryption("Decryption failed".into()))?;

    for keypair in secret_keys {
        let decrypted = msg.decrypt(|| "".to_string(), &[keypair.secret_key()]);

        if let Ok((decrypted_msg, _key_ids)) = decrypted {
            if !verify_with_sender_keys(&decrypted_msg, sender_keys) {
                return Err(CryptoError::Decryption("Decryption failed".into()));
            }

            if let Some(literal) = decrypted_msg.get_literal() {
                return capped_plaintext(literal.data());
            }

            if let Message::Literal(lit) = decrypted_msg {
                return capped_plaintext(lit.data());
            }
        }
    }

    Err(CryptoError::Decryption("Decryption failed".into()))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::encrypt::encrypt_message;
    use crate::keys::generate_keypair;

    #[test]
    fn test_encrypt_decrypt_roundtrip() {
        let keypair = generate_keypair("Test", "test@astermail.com").unwrap();
        let public = keypair.public_key();

        let plaintext = b"This is a secret message that should be encrypted and decrypted!";
        let ciphertext = encrypt_message(plaintext, &[&public]).unwrap();

        let decrypted = decrypt_message(&ciphertext, &[&keypair]).unwrap();

        assert_eq!(plaintext.as_slice(), decrypted.as_slice());
    }

    #[test]
    fn test_decrypt_with_multiple_keys() {
        let alice = generate_keypair("Alice", "alice@astermail.com").unwrap();
        let bob = generate_keypair("Bob", "bob@astermail.com").unwrap();
        let alice_pub = alice.public_key();
        let bob_pub = bob.public_key();

        let plaintext = b"Message for both Alice and Bob";
        let ciphertext = encrypt_message(plaintext, &[&alice_pub, &bob_pub]).unwrap();

        let decrypted_alice = decrypt_message(&ciphertext, &[&alice]).unwrap();
        assert_eq!(plaintext.as_slice(), decrypted_alice.as_slice());

        let decrypted_bob = decrypt_message(&ciphertext, &[&bob]).unwrap();
        assert_eq!(plaintext.as_slice(), decrypted_bob.as_slice());
    }

    #[test]
    fn test_wrong_key_fails() {
        let alice = generate_keypair("Alice", "alice@astermail.com").unwrap();
        let bob = generate_keypair("Bob", "bob@astermail.com").unwrap();
        let alice_pub = alice.public_key();

        let plaintext = b"Secret for Alice only";
        let ciphertext = encrypt_message(plaintext, &[&alice_pub]).unwrap();

        let result = decrypt_message(&ciphertext, &[&bob]);
        assert!(result.is_err());
    }

    fn generate_non_signing_keypair(name: &str, email: &str) -> KeyPair {
        use pgp::composed::{KeyType, SecretKeyParamsBuilder};
        use pgp::crypto::hash::HashAlgorithm;
        use rand::rngs::OsRng;
        use smallvec::smallvec;

        let user_id = format!("{} <{}>", name, email);
        let rng = OsRng;
        let key_params = SecretKeyParamsBuilder::default()
            .key_type(KeyType::Rsa(2048))
            .can_certify(false)
            .can_sign(false)
            .can_encrypt(true)
            .primary_user_id(user_id)
            .preferred_symmetric_algorithms(smallvec![
                pgp::crypto::sym::SymmetricKeyAlgorithm::AES256,
            ])
            .preferred_hash_algorithms(smallvec![HashAlgorithm::SHA2_512])
            .build()
            .unwrap();
        let secret_key = key_params.generate(rng).unwrap();
        let signed_secret_key = secret_key.sign(rng, || "".to_string()).unwrap();
        let armored = signed_secret_key.to_armored_string(None.into()).unwrap();
        crate::keys::import_secret_key(&armored).unwrap()
    }

    #[test]
    fn verified_path_accepts_signing_capable_key() {
        use crate::encrypt::encrypt_and_sign;

        let alice = generate_keypair("Alice", "alice@astermail.com").unwrap();
        let bob = generate_keypair("Bob", "bob@astermail.com").unwrap();
        let alice_pub = alice.public_key();
        let bob_pub = bob.public_key();

        let plaintext = b"signed and encrypted";
        let ciphertext = encrypt_and_sign(plaintext, &[&bob_pub], &alice).unwrap();

        let decrypted = decrypt_and_verify(&ciphertext, &[&bob], &[&alice_pub]).unwrap();
        assert_eq!(plaintext.as_slice(), decrypted.as_slice());
    }

    #[test]
    fn verified_path_rejects_non_signing_sender_key() {
        use crate::encrypt::encrypt_and_sign;

        let signer = generate_non_signing_keypair("NoSign", "nosign@astermail.com");
        let bob = generate_keypair("Bob", "bob@astermail.com").unwrap();
        let signer_pub = signer.public_key();
        let bob_pub = bob.public_key();

        let plaintext = b"forged-by-encryption-only-key";
        let ciphertext = encrypt_and_sign(plaintext, &[&bob_pub], &signer).unwrap();

        let result = decrypt_and_verify(&ciphertext, &[&bob], &[&signer_pub]);
        assert!(result.is_err());
    }
}

File: src/hooks/email_list_helpers/decrypt.ts
//
// Aster Communications Inc.
//
// Copyright (c) 2026 Aster Communications Inc.
//
// This file is part of this project.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the AGPLv3 as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// AGPLv3 for more details.
//
// You should have received a copy of the AGPLv3
// along with this program. If not, see <https://www.gnu.org/licenses/>.
//
import type { DecryptedEnvelope } from "@/types/email";

import {
  get_passphrase_bytes,
  get_passphrase_from_memory,
  get_vault_from_memory,
  wait_for_keys_ready,
} from "@/services/crypto/memory_key_store";
import { decrypt_pgp_message_parallel } from "@/workers/pgp_decrypt_pool";
import {
  adopt_refreshed_vault,
  fetch_refreshed_vault,
} from "@/services/crypto/vault_refresh";
import {
  decrypt_envelope_with_bytes,
  decrypt_envelope_with_identity_key,
  base64_to_array,
  first_base64_byte,
  normalize_envelope_from,
} from "@/services/crypto/envelope";
import { zero_uint8_array } from "@/services/crypto/secure_memory";
import { register_envelope_attachment_keys } from "@/services/crypto/inbound_attachment_keys";
import { decrypt_legacy_ios_envelope } from "@/services/crypto/legacy_ios_envelope";

export async function try_decrypt_with_identity_key(
  encrypted: string | Uint8Array,
  nonce_bytes: Uint8Array,
  identity_key: string,
): Promise<DecryptedEnvelope | null> {
  return decrypt_envelope_with_identity_key(
    identity_key,
    typeof encrypted === "string" ? base64_to_array(encrypted) : encrypted,
    nonce_bytes,
    (plaintext) => {
      const parsed = JSON.parse(new TextDecoder().decode(plaintext));
      const from = normalize_envelope_from(parsed.from);

      if (from) parsed.from = from;

      return parsed;
    },
  );
}

export async function decrypt_envelope(
  encrypted: string,
  nonce: string,
  mail_item_id?: string,
): Promise<DecryptedEnvelope | null> {
  const envelope = await open_envelope(encrypted, nonce, mail_item_id);

  register_envelope_attachment_keys(mail_item_id, envelope);

  return envelope;
}

async function open_envelope(
  encrypted: string,
  nonce: string,
  mail_item_id?: string,
): Promise<DecryptedEnvelope | null> {
  const nonce_bytes = nonce ? base64_to_array(nonce) : new Uint8Array(0);

  if (nonce_bytes.length === 0) {
    try {
      const encrypted_bytes = base64_to_array(encrypted);
      const text = new TextDecoder().decode(encrypted_bytes);

      if (!text.startsWith("-----BEGIN PGP")) {
        return JSON.parse(text) as DecryptedEnvelope;
      }

      let vault = get_vault_from_memory();
      let pass = get_passphrase_from_memory();

      if (!vault?.identity_key || !pass) {
        await wait_for_keys_ready();
        vault = get_vault_from_memory();
        pass = get_passphrase_from_memory();
      }

      if (!vault?.identity_key || !pass) return null;

      const passphrase = pass;
      const decrypt_pgp_with_keys = async (keys: string[]) => {
        const decrypted = await decrypt_pgp_message_parallel(
          text,
          keys,
          passphrase,
        );

        return JSON.parse(decrypted) as DecryptedEnvelope;
      };
      const pgp_keys = [vault.identity_key, ...(vault.previous_keys ?? [])];

      try {
        return await decrypt_pgp_with_keys(pgp_keys);
      } catch (pgp_error) {
        const refreshed = await fetch_refreshed_vault();

        if (refreshed?.vault.identity_key) {
          const tried = new Set(pgp_keys);
          const refreshed_keys = [
            refreshed.vault.identity_key,
            ...(refreshed.vault.previous_keys ?? []),
          ].filter((key) => !tried.has(key));

          if (refreshed_keys.length > 0) {
            const healed = await decrypt_pgp_with_keys(refreshed_keys);

            await adopt_refreshed_vault(refreshed);

            return healed;
          }
        }

        throw pgp_error;
      }
    } catch {
      return null;
    }
  }

  const passphrase = get_passphrase_bytes();

  if (!passphrase) return null;

  try {
    if (nonce_bytes.length === 1 && nonce_bytes[0] === 1) {
      const result = await decrypt_envelope_with_bytes<DecryptedEnvelope>(
        encrypted,
        passphrase,
      );

      zero_uint8_array(passphrase);

      return result;
    }

    zero_uint8_array(passphrase);

    const first_byte = first_base64_byte(encrypted);

    if (
      nonce_bytes.length === 12 &&
      (first_byte === 2 || first_byte === 3 || first_byte === 4)
    ) {
      const { decrypt_mail_envelope } = await import(
        "@/components/email/shared/decrypt_envelope"
      );
      const ecies_result = await decrypt_mail_envelope<DecryptedEnvelope>(
        encrypted,
        nonce,
        mail_item_id,
      );

      if (ecies_result) return ecies_result;
    }

    let vault = get_vault_from_memory();

    if (!vault?.identity_key) {
      await wait_for_keys_ready();
      vault = get_vault_from_memory();
    }

    const encrypted_bytes = base64_to_array(encrypted);

    const try_identity_keys = async (identity_keys: string[]) => {
      for (const identity_key of identity_keys) {
        const decrypted = await try_decrypt_with_identity_key(
          encrypted_bytes,
          nonce_bytes,
          identity_key,
        );

        if (decrypted) return decrypted;
      }

      return null;
    };

    const identity_keys = vault?.identity_key
      ? [vault.identity_key, ...(vault.previous_keys ?? [])]
      : [];
    const result = await try_identity_keys(identity_keys);

    if (result) return result;

    const refreshed = await fetch_refreshed_vault();

    if (refreshed) {
      const tried = new Set(identity_keys);
      const refreshed_keys = [
        ...(refreshed.vault.identity_key ? [refreshed.vault.identity_key] : []),
        ...(refreshed.vault.previous_keys ?? []),
      ].filter((key) => !tried.has(key));

      if (refreshed_keys.length > 0) {
        const healed = await try_identity_keys(refreshed_keys);

        if (healed) {
          await adopt_refreshed_vault(refreshed);

          return healed;
        }
      }
    }

    const legacy_plaintext = await decrypt_legacy_ios_envelope(
      encrypted_bytes,
      nonce_bytes,
    );

    if (legacy_plaintext) {
      const parsed = JSON.parse(new TextDecoder().decode(legacy_plaintext));
      const legacy_from = normalize_envelope_from(parsed.from);

      if (legacy_from) parsed.from = legacy_from;

      return parsed as DecryptedEnvelope;
    }

    return null;
  } catch {
    zero_uint8_array(passphrase);

    return null;
  }
}
